listenbrainz/bigquery/__init__.py

```diff
@@ -0,0 +1,207 @@
+import googleapiclient
+import logging
+import os
+import time
+
+from googleapiclient import discovery
+from googleapiclient.errors import HttpError
+from listenbrainz import config
+from oauth2client.client import GoogleCredentials
+
+APP_CREDENTIALS_FILE = os.environ.get('GOOGLE_APPLICATION_CREDENTIALS')
+JOB_COMPLETION_CHECK_DELAY = 5
+DML_STREAMING_ROWS_DELAY = 30
+
+logger = logging.getLogger(__name__)
+logger.setLevel(logging.INFO)
+
+def create_bigquery_object():
+    """ Initiates the connection to Google BigQuery. Returns a BigQuery object. """
+
+    if not APP_CREDENTIALS_FILE:
+        logger.error("The GOOGLE_APPLICATIONS_CREDENTIALS variable is undefined, cannot connect to BigQuery")
+        raise NoCredentialsVariableException
+
+    if not os.path.exists(APP_CREDENTIALS_FILE):
+        logger.error("The BigQuery credentials file does not exist, cannot connect to BigQuery")
+        raise NoCredentialsFileException
+
+    credentials = GoogleCredentials.get_application_default()
+    return discovery.build('bigquery', 'v2', credentials=credentials)
+
+
+def get_parameters_dict(parameters):
+    """ Converts a list of parameters to be passed to BigQuery into the standard format that the API requires.
+        The format can be seen here:
+        https://developers.google.com/resources/api-libraries/documentation/bigquery/v2/python/latest/bigquery_v2.jobs.html#query
+
+        Args: parameters: a list of dictionaries of the following form
+                {
+                    "name" (str): name of the parameter,
+                    "type" (str): type of the parameter,
+                    "value" (str): value of the parameter
+                }
+
+        Returns: A list of dictionaries that can be passed to the API call
+    """
+
+    if not parameters:
+        return None
+
+    bq_params = []
+    for param in parameters:
+        # construct parameter dict
+        temp = {}
+        temp["name"] = param["name"]
+        temp["parameterType"] = {
+            "type": param["type"],
+        }
+        temp["parameterValue"] = {
+            "value": param["value"],
+        }
+
+        # append parameter dict to main list
+        bq_params.append(temp)
+
+    return bq_params
+
+
+def wait_for_completion(projectId, jobId):
+    """ Make requests periodically until the passed job has been completed """
+
+    while True:
+
+        try:
+            job = bigquery.jobs().get(projectId=projectId, jobId=jobId).execute(num_retries=5)
```

Comment:
What error exactly is it throwing? When I tried, it deleted ~60000 listens almost instantly. But the table was small so maybe that's why.

Reply:
I'm not talking about a concrete situation, but a hypothetical one -- in case BQ is busy or shitty net connections between Hetzner and Google cloud. What do we do if this trips up?

Reply:
It waits until the connection comes back and it gets a response that the job is completed.